File: 0062-unique-paths/0062-unique-paths.java
class Solution {
    public static int f(int[][] dp,int m,int n){
        if(m==0 && n==0) return 1;
        if(n<0 || m<0) return 0;
        if(dp[m][n]!=-1) return dp[m][n];
        return dp[m][n]=f(dp,m-1,n)+f(dp,m,n-1);
    }
    public int uniquePaths(int m, int n) {
        int[][] dp=new int[m][n];
        for(int[] x:dp){
            Arrays.fill(x,-1);
        }

        return f(dp,m-1,n-1);
    }
}

File: 0062-unique-paths/0062-unique-paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[-1]*n for _ in range(m)]
        def f(dp, m, n):
            if n == 0 and m == 0:
                return 1
            if m < 0 or n < 0:
                return 0
            if dp[m][n] != -1:
                return dp[m][n]
            
            dp[m][n] = f(dp, m-1, n) + f(dp, m, n-1)
            return dp[m][n]

        return f(dp, m-1, n-1)